da/celestia/celestia.go

```diff
@@ -159,62 +174,63 @@ func (c *DataAvailabilityLayerClient) SubmitBatch(batch *types.Batch) da.ResultS
-									DAHeight: inclusionHeight,
-								},
-							}
-						}
-					} else {
-						c.logger.Debug("Failed to receive DA batch inclusion result. Emitting health event and trying again", "error", err)
-						res, err := da.SubmitBatchHealthEventHelper(c.pubsubServer, c.ctx, false, err)
-						if err != nil {
-							return res
-						}
-					}
-
-				} else {
-					c.logger.Debug("Successfully submitted DA batch", "txResponse", txResponse.RawLog)
-					res, err := da.SubmitBatchHealthEventHelper(c.pubsubServer, c.ctx, true, nil)
-					if err != nil {
-						return res
-					}
-					return da.ResultSubmitBatch{
-						BaseResult: da.BaseResult{
-							Code:     da.StatusSuccess,
-							Message:  "tx hash: " + txResponse.TxHash,
-							DAHeight: uint64(txResponse.Height),
-						},
-					}
+			estimatedGas := float64(EstimateGas([]int{len(blob)}))
+			gasWanted := uint64(estimatedGas * gasAdjustment)
+			fees := c.calculateFees(gasWanted)
+
+			//SubmitPFB sets an error if the txResponse has error, so we check check the txResponse for error
+			txResponse, err := c.client.SubmitPFB(c.ctx, c.config.NamespaceID, blob, int64(fees), gasWanted)
+			if txResponse == nil {
+				c.logger.Debug("Failed to submit DA batch. Emitting health event and trying again", "error", err)
+				// Publish an health event. Only if we failed to emit the event we return an error.
+				res, err := da.SubmitBatchHealthEventHelper(c.pubsubServer, c.ctx, false, err)
+				if err != nil {
+					return res
 				}
-			} else {
-				res, err := da.SubmitBatchHealthEventHelper(c.pubsubServer, c.ctx, false, errors.New("DA txResponse is nil"))
+				time.Sleep(c.submitRetryDelay)
+				continue
+			}
+
+			if txResponse.Code != 0 {
+				c.logger.Debug("Failed to submit DA batch. Emitting health event and trying again", "txResponse", txResponse.RawLog, "code", txResponse.Code)
+				// Publish an health event. Only if we failed to emit the event we return an error.
+				res, err := da.SubmitBatchHealthEventHelper(c.pubsubServer, c.ctx, false, errors.New(txResponse.RawLog))
 				if err != nil {
 					return res
 				}
 				time.Sleep(c.submitRetryDelay)
+				continue
+			}
+
+			// Here we assume that if txResponse is not nil and also error is not nil it means that the transaction
+			// was submitted (not necessarily accepted) and we still didn't get a clear status regarding it (e.g timeout).
+			// hence trying to poll for it.
+			daHeight := uint64(txResponse.Height)
+			if daHeight == 0 {
+				c.logger.Debug("Failed to receive DA batch inclusion result. Waiting for inclusion", "txHash", txResponse.TxHash)
+				daHeight, err = c.waitForTXInclusion(txResponse.TxHash)
+				if err != nil {
+					c.logger.Debug("Failed to receive DA batch inclusion result. Emitting health event and trying again", "error", err)
+					res, err := da.SubmitBatchHealthEventHelper(c.pubsubServer, c.ctx, false, err)
+					if err != nil {
+						return res
+					}
+					time.Sleep(c.submitRetryDelay)
```

Comment:
possibly all of this can be exported to a function which gets the error as an argument as this logic repeats in the first 3 if statements.